### PR TITLE
feat: table component for custom dashboard

### PR DESCRIPTION
**Pull request by djbarnwal**
Closes [#4851](https://github.com/rilldata/rill/issues/4851)

Spec 

```
kind: component

table:
  metric_view: Bids_Sample_Dash
  time_range: P7D
  measures:
    - measure_2
  row_dimensions:
    - player_size
  col_dimensions: 
    - app_or_site
```

image

### CODE REVIEW

**Comment by briangregoryholmes** [comment]
Application crashed for me on this uncaught error. Just copied from your spec, but, with auto save, didn't get a chance to change the metric view and dimension/measure names to ones that were valid for my project.

Screenshot 2024-07-09 at 11 01 57 AM

**Comment by briangregoryholmes** [comment]
More specific error when the measure is invalid:

Screenshot 2024-07-09 at 11 08 34 AM

**Comment by nishantmonu51** [comment]
@djbarnwal : Is this ready to be merged ?

**Comment by briangregoryholmes** [comment]
https://jam.dev/c/f81d00f5-834b-4488-8241-a3a1873ff3ea

Still having a few issues. Changing the dimension name to another valid dimension doesn't update the table when saving. In this recording, it required two refreshes, though that isn't the case in every browser.

With auto-save turned on, changing something like metric view name to something that isn't valid no longer hangs the application, but it does require a refresh of the page once you change it back to something valid.

File: runtime/compilers/rillv1/data/component-template-v1.json [hunk 1]
```diff
@@ -145,10 +148,8 @@
         }
       },
       "required": [
-        "col_dimensions",
         "measures",
         "metric_view",
```
Comment (ericpgreen2) [inline runtime/compilers/rillv1/data/component-template-v1.json:152]: Nit: we should consolidate on "metrics_view" throughout the codebase, not "metric_view"

File: web-common/src/features/templates/table/TableTemplate.svelte [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  $: if (isValidSchema) {
+    const stateManagerContext = createStateManagers({
+      queryClient,
+      metricsViewName: tableProperties.metric_view,
+      extraKeyPrefix: TABLE_PREFIX,
+    });
+
+    pivotConfig = getTableConfig(instanceId, tableProperties, $pivotState);
+    pivotDataStore = createPivotDataStore(stateManagerContext, pivotConfig);
+  }
```
Comment (ericpgreen2) [inline web-common/src/features/templates/table/TableTemplate.svelte:74]: Writing down what we discussed [in Slack](https://rilldata.slack.com/archives/C04AGHL77PS/p1722346431806149?thread_ts=1722321321.527189&cid=C04AGHL77PS) – we likely shouldn't re-create the `pivotDataStore` every time the `tableProperties` change, but rather we should update the store's values.

**Review by ericpgreen2 — APPROVED** [review]
The table editing experience is good! Great component to have.

(UXQA) The table component renders oddly in the Custom Dashboard workspace. It's small and doesn't fill the space of the container:
image

File: web-common/src/features/templates/table/TableTemplate.svelte (hunk 2 above)
Comment (djbarnwal) [reply web-common/src/features/templates/table/TableTemplate.svelte:74]: Will address in follow up

**Comment by djbarnwal** [comment]
> (UXQA) The table component renders oddly in the Custom Dashboard workspace. It's small and doesn't fill the space of the container

Sizing in general is an issue which needs to be worked out for all components. It's tricky because we don't want to always increase the font size based on container size. This needs more discussion offline.

File: runtime/compilers/rillv1/data/component-template-v1.json (hunk 1 above)
Comment (djbarnwal) [reply runtime/compilers/rillv1/data/component-template-v1.json:152]: This is a common change for all components. Will create a separate PR.

**Pull request merged**